### ISSUE TITLE
Pull To Refresh feature in Webview

### ISSUE THREAD

**Issue by ankur-sardar** [react-native-community/react-native-webview#103]
Is there any good way to support something like `refreshControl`? 
I saw one discussion regarding this - https://github.com/facebook/react-native/issues/7870

**Comment by Titozzz** [comment]
Maybe it would be possible but I'm not sure you want to do that now as there are some issue of stuff that only happens on first render such as injectedJS and headers sent and stuff

**Comment by codezz** [comment]
Hi, I managed to get it working using  Refresh Control on a wrapping ScrollView and calling webview.relod() when doing refresh like suggested here: 
https://github.com/facebook/react-native/issues/7870#issuecomment-292767858

**Comment by ankur-sardar** [comment]
@codezz  This solution has some demerits, also this solution is not good for android webview.

**Comment by Titozzz** [comment]
Yeah I don't think using a component around would be a good idea. We might want to detect gestures ourselves or something

**Comment by ankur-sardar** [comment]
May be we can use `SwipeRefreshLayout` for Android and `UIRefreshControl` for ios. I am able to add the solution for iOS, but I can not make any solution for the android webview. (I don't have that much Android experience). I can not understand how can I add SwipeRefreshLayout in webview manager.

**Comment by jermaine** [comment]
@ankur-sardar Have you had the time to implement the solution for iOS? Was looking through your fork and came across this: [commit-9049d02e](https://github.com/ankur-sardar/react-native-webview/commit/9049d02e65ec2adae9d2d4ecf146c49f0df74d67
), which looks like a work in progress...

I do think it's important to have the RefreshControl be managed directly by the WebView instead of wrapping it with a ScrollView.

**Comment by ankur-sardar** [comment]
@jermaine  I have implemented for iOS version, in the Native layer. iOS version I used this approach - https://github.com/ankur-sardar/react-native-wkwebview/commit/b3577ab5e5ca30b5bd3685466363a1ebe7b7ae02
It is working fine. But the main issue I am facing is in Android WebView. It seems CloneElement `RefreshControl` ([SwipeRefreshLayout](https://developer.android.com/reference/android/support/v4/widget/SwipeRefreshLayout)) with the Webview is not a good solution. in some websites (google Photos) whenever I try to scroll up, it triggers the `SwipeRefreshLayout` , I found the reason behind this, but could not get any good solution for this yet.

**Comment by yulolimum** [comment]
For anyone looking for a workaround, here's a relatively complete example based on this: https://github.com/facebook/react-native/issues/7870#issuecomment-292767858

``` typescript
import * as React from "react"
import { WebView as RNWebView } from "react-native-webview"
import { RefreshControl, ViewStyle, ScrollView } from "react-native"

// prettier-ignore
const INJECTED_JS = `
  window.onscroll = function() {
    window.ReactNativeWebView.postMessage(
      JSON.stringify({
        scrollTop: document.documentElement.scrollTop || document.body.scrollTop
      }),     
    )
  }
`

interface ExampleState {
  isPullToRefreshEnabled: boolean
  scrollViewHeight: number
}

interface ExampleProps {}

const SCROLLVIEW_CONTAINER: ViewStyle = {
  flex: 1,
  height: "100%",
}

const WEBVIEW = (height): ViewStyle => ({
  width: "100%",
  height,
})

export class Example extends React.Component<ExampleProps, ExampleState> {
  webView: any

  state = {
    isPullToRefreshEnabled: false,
    scrollViewHeight: 0,
  }

  setWebViewRef = ref => {
    this.webView = ref
  }

  onRefresh = () => this.webView.reload()

  onWebViewMessage = e => {
    const { data } = e.nativeEvent

    try {
      const { scrollTop } = JSON.parse(data)
      this.setState({ isPullToRefreshEnabled: scrollTop === 0 })
    } catch (error) {}
  }

  render() {
    const { scrollViewHeight, isPullToRefreshEnabled } = this.state

    return (
      <ScrollView
        style={SCROLLVIEW_CONTAINER}
        onLayout={e => this.setState({ scrollViewHeight: e.nativeEvent.layout.height })}
        refreshControl={
          <RefreshControl
            refreshing={false}
            enabled={isPullToRefreshEnabled}
            onRefresh={this.onRefresh}
            tintColor="transparent"
            colors={["transparent"]}
            style={{ backgroundColor: "transparent" }}
          />
        }>
        <RNWebView
          source={{ uri: "https://infinite.red" }}
          ref={this.setWebViewRef}
          style={WEBVIEW(scrollViewHeight)}
          onMessage={this.onWebViewMessage}
          injectedJavaScript={INJECTED_JS}
        />
      </ScrollView>
    )
  }
}

```

**Comment by jamonholmgren** [comment]
Let's add that to our docs, @yulolimum , really cool!

**Comment by GammelBro** [comment]
This solution works great but on ios devices if you pull down and rapidly scroll down without triggering the pull to refresh function you cant scroll in the webview. Did someone have a fix for that?

**Comment by philipp65** [comment]
Since `injectedJavaScript` combined with `window.ReactNativeWebView.postMessage` has a significant delay, this was not an option for me.

I started off by forking and implementing the `PTR` native, based on https://github.com/react-native-community/react-native-webview/pull/362. I managed to get it working on both Android and iOS (without documentation and some useful bridges for coloring, manual enable and global disable of `PTR`,..) but then discontinued because I wanted to have the same _RefreshView_ if the page is loading.

Long story short. I discovered an undocumented bridge in this `WebView` including the native `onScroll` events. With this information I built both Android and iOS with the Material Spinner.

If someone is interested, I could look through the code and provide some snippets.

**Comment by jackmu95** [comment]
@philipp65 Would be nice if you could share your code 👍

**Comment by philipp65** [comment]
I tried to shrink the code as good as possible. On iOS things work with less code since the `WebView` bounces and we can determine when the user has reached the top and how far "over the top" the user has scrolled. For Android I used custom `Responder` events.

`MyWebview.ios.js`
```js
...

export default class MyWebView extends MyBaseWebView {
    offsetPTR = 0;
    firstPTR = true;
    scrollEnabled = true;

    constructor(props) {
        super(props);
        this.state = {
            scrollEnabled: true,
            ...      
        };
    }

    ...

    onScroll = (navState) => {
        if (navState.nativeEvent.contentOffset.y < -5) {
            if (this.firstPTR) {
                this.offsetPTR = navState.nativeEvent.contentOffset.y;
                this.firstPTR = false;
            } else {
                if (this.offsetPTR - navState.nativeEvent.contentOffset.y > 0) {
                    this.setState({
                        progressTopMarginOffset: this.offsetPTR - navState.nativeEvent.contentOffset.y,
                    });
                }

                if (this.offsetPTR - navState.nativeEvent.contentOffset.y > 100) {
                    this.firstPTR = true;
                    this.handleReload();
                    this.setState({
                        progressTopMarginOffset: 100,
                        progress: 0,
                        scrollEnabled: false
                    }, this.enableScroll);
                }
            }

        } else {
            this.firstPTR = true;
            this.setState({
                progressTopMarginOffset: 100,
            });
        }
    };

    enableScroll = () => {
        this.setState({
            scrollEnabled: true,
        });
    };

    render() {
        return (
            <>
                <WebView

                    ...

                    scrollEnabled={this.state.scrollEnabled}
                    onScroll={this.onScroll}

                />

                {((this.state.progress < 0.99 && !this.state.loaded) || this.state.progressTopMarginOffset !== 100) &&
                <View style={{...styles.progress_container, top: 60 + this.state.progressTopMarginOffset, opacity: this.state.progressTopMarginOffset / 100}}>
                    <TouchableOpacity style={styles.progress}>
                        <ActivityIndicator size="small" color="#fff"/>
                    </TouchableOpacity>
                </View>
                }

                {this.state.epaperUrl !== "" &&
                <ModalEPaper setEpaperUrl={this.setEpaperUrl} openNative={this.openNative} epaperUrl={this.state.epaperUrl}/>
                }
            </>
        );
    }
}

const styles = StyleSheet.create({
    ...
});

```

`MyWebView.android.js`
```js
...

export default class MyWebView extends MyBaseWebView {
    offsetPTR = 0;
    firstPTR = true;
    scrollEnabled = true;

    constructor(props) {
        super(props);
        this.state = {
            scrollEnabled: true,
            ...      
        };
    }

    ...

    onScroll = (navState) => {
        this.enablePTR = navState.nativeEvent.contentOffset.y <= 1
    };

    onTouchEvent = (name, navState) => {
        if (this.enablePTR) {
            if (this.firstPTR) {
                this.offsetPTR = navState.nativeEvent.pageY;
                this.firstPTR = false;
            } else {
                if (this.offsetPTR < 300) {
                    this.setState({
                        progressTopMarginOffset: navState.nativeEvent.pageY - this.offsetPTR,
                    });
                }
                if (navState.nativeEvent.pageY - this.offsetPTR > 100) {
                    this.offsetPTR += 10000;
                    this.handleReload();
                    this.setState({
                        progressTopMarginOffset: 100,
                        progress: 0,
                    });
                }
            }
            if (name === 'onResponderRelease') {
                this.firstPTR = true;
                this.setState({
                    progressTopMarginOffset: 100,
                });
            }
        } else {
            this.firstPTR = true;
            this.setState({
                progressTopMarginOffset: 100,
            });
        }
    };

    enableScroll = () => {
        this.setState({
            scrollEnabled: true,
        });
    };

    render() {
        return (
            <>
                <View style={{width: '100%', height: '100%'}}
                      onStartShouldSetResponder={(ev) => true}
                      onMoveShouldSetResponder={(ev) => true}
                      onResponderMove={this.onTouchEvent.bind(this, "onResponderMove")}
                      onResponderRelease={this.onTouchEvent.bind(this, "onResponderRelease")}
                      onResponderTerminationRequest={(ev) => true}
                >
                    <WebView

                        ...

                        scrollEnabled={this.state.scrollEnabled}
                        onScroll={this.onScroll}

                    />

                    {((this.state.progress < 0.99 && !this.state.loaded) || this.state.progressTopMarginOffset !== 100) &&
                    <View style={{...styles.progress_container, top: 60 + this.state.progressTopMarginOffset, opacity: this.state.progressTopMarginOffset / 100}}>
                        <TouchableOpacity style={styles.progress}>
                            <ActivityIndicator size="small" color="#fff"/>
                        </TouchableOpacity>
                    </View>
                    }

                    {this.state.epaperUrl !== "" &&
                    <ModalEPaper setEpaperUrl={this.setEpaperUrl} openNative={this.openNative} epaperUrl={this.state.epaperUrl}/>
                    }
                </View>
            </>
        );
    }
}

const styles = StyleSheet.create({
    ...
});
```

Let me know if there is any major part missing.

**Comment by xclidongbo** [comment]
@yulolimum  It is not work for the first time。

**Comment by hari** [comment]
For anyone coming here, I just created this and it is working for me.

```js
import React, { useState } from 'react';
import WebView from 'react-native-webview';
import { RefreshControl, Dimensions, StyleSheet } from 'react-native';
import { ScrollView } from 'react-native-gesture-handler';

const styles = StyleSheet.create({
  view: { flex: 1, height: '100%' }
});

const RefreshWebView = ({ isRefresh, onRefresh, ...webViewProps }) => {
  const [height, setHeight] = useState(Dimensions.get('screen').height);
  const [isEnabled, setEnabled] = useState(typeof onRefresh === 'function');
  return (
    <ScrollView
      onLayout={(e) => setHeight(e.nativeEvent.layout.height)}
      refreshControl={
        <RefreshControl
          onRefresh={onRefresh}
          refreshing={isRefresh}
          enabled={isEnabled}
        />
      }
      style={styles.view}>
      <WebView
        {...webViewProps}
        onScroll={(e) =>
          setEnabled(
            typeof onRefresh === 'function' &&
              e.nativeEvent.contentOffset.y === 0
          )
        }
        style={[styles.view, { height }, webViewProps.style]}
      />
    </ScrollView>
  );
};

export default RefreshWebView;
```

**Comment by panda0603** [comment]
@hari Thanks, though does this work properly in google maps, or any websites with similar embedded scrollable view?

**Comment by ohsik** [comment]
@hari Thank you for sharing this. It works great on iOS and Android!!

**Comment by neel2292** [comment]
@hari this doesnt work well for me, as I have elements in my webview with height set to 100%. Hence, what happened when scrolling through aforementioned element is the onscroll event doesn't get fired. Hence, the pull to refresh is fired on every scroll down. My solution to this is basically listening to the touchmove event in script injection and get the current scrollTop to determine whether the pull to refresh should be enabled

```
var prevTop = 0,
	currentTop = 0,
	scrollTimeout;

window.addEventListener('touchmove', function (e) {
	clearTimeout(scrollTimeout);
	scrollTimeout = setTimeout(function (){
		handleTouchMove(e);
	}, 100);
}, false);

function handleTouchMove (e) {
	currentTop = getScrollTop(e.target);

	if (prevTop !== currentTop && prevTop === 0 && currentTop > 0) {
		window.WebViewBridge.send('DisableRefresh');
	}

	if (prevTop !== currentTop && currentTop === 0) {
		window.WebViewBridge.send('EnableRefresh');
	}

	prevTop = currentTop;

	function getScrollTop(target) {
		while (target && target !== document && target.scrollTop === 0) {
			target = target.parentElement;
		}

		return (target ? target.scrollTop : 0) || document.documentElement.scrollTop || document.body.scrollTop;
	}
}
```

**Comment by cesarfavero** [comment]
> @hari Thank you for sharing this. It works great on iOS and Android!!

Dont work in android my friend...

**Comment by muskan273** [comment]
Hi I tried to implement pull to refresh webview with android native code and it's working for me.
you can use the below package to enable refresh with webview on android  **https://github.com/muskan273/react-native-webview**

**Comment by HerickRaposo** [comment]
as I said before,  i tried to implement it this way, however I can only scroll down but when I need to scroll back it gives refrash, however I didn't get good results because the RefreshControl was always enabled, so I made the following adjustments along with the previously mentioned solutions:


```
const [refreshing, setRefreshing] = useState(false);
const [refresherEnabled, setEnableRefresher] = useState(true);


  //Code to get scroll position
  const handleScroll = (event) =>  {
    console.log(Number(event.nativeEvent.contentOffset.y))
    const yOffset = Number(event.nativeEvent.contentOffset.y)
    if (yOffset === 0){
      console.log('top of the page')
      setEnableRefresher(true)
    }else{
      setEnableRefresher(false)
    }
  }
    
return (
    <SafeAreaView style={genericStyles.containerWebView}>
      <StatusBar backgroundColor={genericStyles.colors.BlueBg} barStyle={'light-content'} />
      <ScrollView style={genericStyles.scrollView} 
      contentContainerStyle={{flex: 1}}
      refreshControl={
        <RefreshControl 
          refreshing={refreshing}
          enabled={refresherEnabled}
          onRefresh={()=>{webViewRef.current.reload()}}
        />
      }>
        <WebView
          source={{ uri: your url }}
          ref={webViewRef}
          originWhitelist={['*']}
          allowsInlineMediaPlayback
          javaScriptEnabled
          scalesPageToFit
          mediaPlaybackRequiresUserAction={false}
          javaScriptEnabledAndroid
          useWebkit
          startInLoadingState={true}
          cacheEnabled
          onScroll={handleScroll}
        />
      </ScrollView>
    </SafeAreaView>
  )
}
```

**Comment by eliiran1231** [comment]
as exstention to @HerickRaposo's sulotion i figured you could rewrite the handleScroll function to this: 

` 
    const handleScroll = (event: { nativeEvent: { contentOffset: { y: any; }; }; }) =>  {
    const yOffset = Number(event.nativeEvent.contentOffset.y)
    if (yOffset === 0){
      setEnableRefresher(true);
    }
    else if(refresherEnabled) setEnableRefresher(false);
  }
`
add this to onRefresh of the RefreshControl:
`
setRefreshing(true);
`
and add this to the webview: 
`
onLoadEnd={()=>setRefreshing(false)}
`
end result on the code example would look something like this:

`
const [refreshing, setRefreshing] = useState(false);
const [refresherEnabled, setEnableRefresher] = useState(true);


  //Code to get scroll position
  const handleScroll = (event) =>  {
    console.log(Number(event.nativeEvent.contentOffset.y))
    const yOffset = Number(event.nativeEvent.contentOffset.y)
    if (yOffset === 0){
      console.log('top of the page')
      setEnableRefresher(true)
    }else if(refresherEnabled){
      setEnableRefresher(false)
    }
  }
    
return (
    <SafeAreaView style={genericStyles.containerWebView}>
      <StatusBar backgroundColor={genericStyles.colors.BlueBg} barStyle={'light-content'} />
      <ScrollView style={genericStyles.scrollView} 
      contentContainerStyle={{flex: 1}}
      refreshControl={
        <RefreshControl 
          refreshing={refreshing}
          enabled={refresherEnabled}
          onRefresh={()=>{
            webViewRef.current.reload()
            setRefreshing(true);
        }}
        />
      }>
        <WebView
          source={{ uri: "your url" }}
          ref={webViewRef}
          originWhitelist={['*']}
          allowsInlineMediaPlayback
          javaScriptEnabled
          scalesPageToFit
          mediaPlaybackRequiresUserAction={false}
          javaScriptEnabledAndroid
          useWebkit
          startInLoadingState={true}
          cacheEnabled
          onScroll={handleScroll}
          onLoadEnd={setRefreshing(false)}
        />
      </ScrollView>
    </SafeAreaView>
  )
`

should be slightly more efficiant, and the refresh will stay untill the page finish its load, if that is what you are looking for!

**Comment by Al-homaidi** [comment]
this is work if soomone need it 

import React, { useState, useRef } from 'react';
import { SafeAreaView, ScrollView, RefreshControl, StatusBar, NativeSyntheticEvent, NativeScrollEvent } from 'react-native';
import WebView from 'react-native-webview';

const App = () => {
  const [refreshing, setRefreshing] = useState(false);
  const [refresherEnabled, setEnableRefresher] = useState(false);
  const webViewRef = useRef(null);

  const handleScroll = (event: NativeSyntheticEvent<NativeScrollEvent>) => {
    const yOffset = Number(event.nativeEvent.contentOffset.y);
    console.log('WebView yOffset:', yOffset);

    if (yOffset <= 1) {
      console.log('Top of the page inside WebView');
      setEnableRefresher(true);
    } else {
      setEnableRefresher(false);
    }
  };

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: '#fff' }}>
      <StatusBar backgroundColor="#007bff" barStyle="light-content" />
      <ScrollView
        style={{ flex: 1 }}
        contentContainerStyle={{ flexGrow: 1 }}
        scrollEnabled={false}
        refreshControl={
          <RefreshControl
            refreshing={refreshing}
            enabled={refresherEnabled}
            onRefresh={() => {
              webViewRef.current?.reload();
              setRefreshing(true);
              setTimeout(() => setRefreshing(false), 1000);
            }}
          />
        }
      >
        <WebView
          source={{ uri: 'https://bjda.org' }}
          ref={webViewRef}
          originWhitelist={['*']}
          allowsInlineMediaPlayback
          javaScriptEnabled
          scalesPageToFit
          mediaPlaybackRequiresUserAction={false}
          useWebKit
          startInLoadingState={true}
          cacheEnabled
          onScroll={handleScroll}
          onLoadEnd={() => setRefreshing(false)}
        />
      </ScrollView>
    </SafeAreaView>
  );
};

export default App;

**Comment by redkean** [comment]
> [@hari](https://github.com/hari) this doesnt work well for me, as I have elements in my webview with height set to 100%. Hence, what happened when scrolling through aforementioned element is the onscroll event doesn't get fired. Hence, the pull to refresh is fired on every scroll down. My solution to this is basically listening to the touchmove event in script injection and get the current scrollTop to determine whether the pull to refresh should be enabled
> 
> var prevTop = 0,
> 	currentTop = 0,
> 	scrollTimeout;
> 
> window.addEventListener('touchmove', function (e) {
> 	clearTimeout(scrollTimeout);
> 	scrollTimeout = setTimeout(function (){
> 		handleTouchMove(e);
> 	}, 100);
> }, false);
> 
> function handleTouchMove (e) {
> 	currentTop = getScrollTop(e.target);
> 
> 	if (prevTop !== currentTop && prevTop === 0 && currentTop > 0) {
> 		console.log('DisableRefresh');
> 	}
> 
> 	if (prevTop !== currentTop && currentTop === 0) {
> 		console.log('EnableRefresh');
> 	}
> 
> 	prevTop = currentTop;
> 
> 	function getScrollTop(target) {
> 		while (target && target !== document && target.scrollTop === 0) {
> 			target = target.parentElement;
> 		}
> 
> 		return (target ? target.scrollTop : 0) || document.documentElement.scrollTop || document.body.scrollTop;
> 	}
> }

this works wonders, thanks!

i'm going to share the basic setup for this later.

**Comment by dev-aly3n** [comment]
> [@hari](https://github.com/hari) this doesnt work well for me, as I have elements in my webview with height set to 100%. Hence, what happened when scrolling through aforementioned element is the onscroll event doesn't get fired. Hence, the pull to refresh is fired on every scroll down. My solution to this is basically listening to the touchmove event in script injection and get the current scrollTop to determine whether the pull to refresh should be enabled
> 
Interesting solution, @neel-actual! But I can imagine an edge case for it. What if the page doesn't have any scroll height (so scrollTop is already 0), and there's a select box on the page? Imagine tapping the select box, which has its own scroll. You scroll down inside the select box and then scroll up. In this case, the page tries to refresh, but the user only intended to scroll up inside the select box.

Do you have any ideas for handling this edge case while using your approach?

**Comment by aivnaov123** [comment]
Is there any good solution already, pullToRefreshEnabled, not working on android devices, using RefreshControl didnt work properly on android devices when there are two nested scrolls? If anyone find a way to handle it please share a solution :)
